ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1421 @@
+### E2E
+
+-->
+
+We should verify the following behavior in tests:
+1. we can continue processing after a restart (i.e. the processing of a command produces the same state as replaying its corresponding events)
+2. we can continue processing after an update (i.e. the events of a previous Zeebe version can be replayed)
+3. we can continue processing after an update with behavior changes (i.e. the processor can handle different behaviors/versions, and the replay of events leads to the same state before and after the update)
+
+The effect of the workflow processing (throughput + latency) and the startup time should be measured in different benchmarks. The assumption is that the processing performance and the startup time improve, or are at least on the same level.  
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+<!--
+Why should we *not* do this?
+-->
+
+The processing of a command produces more records compared to the processing of the single event before. Since all records are written in a batch to ensure an atomic operation, it is more likely to exceed the limit of the message size.   
+
+The new processing produces also more records in total. So, more data is written to the stream and replicated to the followers. 
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+<!--
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+-->
+
+The following section describe different alternatives and their drawbacks.
+
+## New Keys for Records Only
+
+Focusing on the problem of the key generator, an idea is to avoid that the data is written in the state with different keys. Instead of generating a new key for a new entity in the state, the new key is used for writing a record. The new data is written to the state when the record is processed (e.g. generate a new key and write an `ELEMENT_ACTIVATING` record but don't insert the new element instance in the state until reading the record).
+
+This idea doesn't work in cases when the new data is required for processing the next record. For example, two variable update commands for the same variable name and scope. If the second command is processed before the follow-up variable event of the first command is read then it writes a variable event with intent `CREATED` instead of `UPDATED`.
+
+Another example is the processing of a multi-instance activity. For each element in the input collection, an `ELEMENT_ACTIVATING` event is written and the current element is stored as a variable. The variable event must be written after the event `ELEMENT_ACTIVATING` is read to know the scope key of the variable. But the variable is required on processing `ELEMENT_ACTIVATING` to apply the input mappings.
+    
+The idea doesn't cover other behavior changes that can lead to an inconsistent state.     
+    
+## State Changes on Events Only
```

Comment:
Don't really understand this section. In what way does it differ from the idea to reuse the apply logic during replay also in the processing phase? (I am referring to the second paragraph in Reference-level explanation: "The state changes are written when processing a command ... the implementation of these parts should be coupled or reused."

To me it sounds very much like the idea expressed here. 

But I guess here there are different assumptions about ordering of records and thus order of state changes? 

In any case, it would help to explain how this section is different from the concept discussed in the beginning of the reference level explanation.

Reply:
I'll try to improve the description :writing_hand: 

The difference to the proposal is the time when the state changes are applied. Apply the changes when writing the event vs when reading the event.